test/bilinearinterpolation.jl

```diff
@@ -0,0 +1,160 @@
+        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
+        targetmesh = (
+            x = DA([xtargetcpu[i] for i in 1:ntargetx, j in 1:ntargety]),
+            y = DA([ytargetcpu[j] for i in 1:ntargetx, j in 1:ntargety]),
+        )
+        fsource = testfunc.(sourcemesh.x, sourcemesh.y)
+        ftargetexact = testfunc.(targetmesh.x, targetmesh.y)
+        # allocate ftarget
+        ftarget = DA{FT}(undef, ntargetx, ntargety)
+        # use bilinear interpolation
+        interpolatebilinear!(ftarget, bilinear, fsource)
+        l∞error = maximum(abs.((ftarget .- ftargetexact)[:]))
+        @test l∞error ≤ toler
+    end
+    return (trial_constructor, trial_interpolator)
+end
+
+function test_multilevel(
+    ::Type{DA},
+    ::Type{FT},
+    (xmin, xmax),
+    (ymin, ymax),
+    (zmin, zmax),
+    (nsourcex, nsourcey),
+    (ntargetx, ntargety),
+    nlevels;
+    toler,
+) where {DA, FT}
+    @testset "bilinear interpolation on $nlevels levels with $FT, nsource = ($nsourcex, $nsourcey), ntarget = ($ntargetx, $ntargety)" begin
+        # build source and target mesh
+        xsource, xtarget = get_uniform_column_grids(
+            DA,
+            FT,
+            xmin,
+            xmax,
+            xmin,
+            xmax,
+            nsourcex,
+            ntargetx,
+        )
+        ysource, ytarget = get_uniform_column_grids(
+            DA,
+            FT,
+            ymin,
+            ymax,
+            ymin,
+            ymax,
+            nsourcey,
+            ntargety,
+        )
+        z, _ = get_uniform_column_grids(
+            DA,
+            FT,
+            zmin,
+            zmax,
+            zmin,
+            zmax,
+            nlevels,
+            nlevels,
+        )
+        bilinear = Bilinear(xsource, ysource, xtarget, ytarget)
+        # build fsource
+        xsourcecpu, ysourcecpu, zcpu = Array(xsource), Array(ysource), Array(z)
+        sourcemesh = (
+            x = DA([
+                xsourcecpu[j] for i in 1:nlevels, j in 1:nsourcex,
+                k in 1:nsourcey
+            ]),
+            y = DA([
+                ysourcecpu[k] for i in 1:nlevels, j in 1:nsourcex,
+                k in 1:nsourcey
+            ]),
+            z = DA([
+                zcpu[i] for i in 1:nlevels, j in 1:nsourcex, k in 1:nsourcey
+            ]),
+        )
+        xtargetcpu, ytargetcpu = Array(xtarget), Array(ytarget)
+        targetmesh = (
+            x = DA([
+                xtargetcpu[j] for i in 1:nlevels, j in 1:ntargetx,
+                k in 1:ntargety
+            ]),
+            y = DA([
+                ytargetcpu[k] for i in 1:nlevels, j in 1:ntargetx,
+                k in 1:ntargety
+            ]),
+            z = DA([
+                zcpu[i] for i in 1:nlevels, j in 1:ntargetx, k in 1:ntargety
+            ]),
+        )
+        fsource = testfunc.(sourcemesh.x, sourcemesh.y, sourcemesh.z)
+        ftargetexact = testfunc.(targetmesh.x, targetmesh.y, targetmesh.z)
```

Comment:
Do we need to reallocate new arrays here? Why not do a single allocation?

Reply:
I split it out for readability! But I can compress it.